include/circt/Scheduling/Scheduler.h

```diff
@@ -0,0 +1,253 @@
+  //===--------------------------------------------------------------------===//
+public:
+  /// Add \p op to the scheduling problem.
+  void addOperation(Operation *op) { ops.insert(op); }
+
+  /// Construct a dependence that distinguishes result and operands indices.
+  Dependence *getOrInsertDependence(Operation *src, unsigned srcIdx,
+                                    Operation *dst, unsigned dstIdx);
+
+  /// Construct a dependence that uses the default result and operand indices.
+  Dependence *getOrInsertDependence(Operation *src, Operation *dst) {
+    return getOrInsertDependence(src, 0, dst, 0);
+  }
+
+  /// Register an operator type identified by \p name.
+  OperatorType *getOrInsertOperatorType(StringRef name);
+
+  //===--------------------------------------------------------------------===//
+  // Subclass access to problem components
+  //===--------------------------------------------------------------------===//
+protected:
+  bool hasOperation(Operation *op) { return ops.contains(op); }
+  const OpSet &getOperations() { return ops; }
+
+  bool hasDependence(Dependence *dep) { return deps.contains(dep); }
+  const DepSet &getDependences() { return deps; }
+
+  bool hasOperatorType(OperatorType *opr) { return oprs.contains(opr); }
+  const OprSet &getOperatorTypes() { return oprs; }
+
+  //===--------------------------------------------------------------------===//
+  // Subclass access to properties: retrieve problem, set solution
+  //===--------------------------------------------------------------------===//
+protected:
+  bool hasAssociatedOperatorType(Operation *op) { return assocOpr.count(op); }
+  OperatorType *getAssociatedOperatorType(Operation *op) {
+    return assocOpr.lookup(op);
+  }
+
+  bool hasLatency(OperatorType *opr) { return latency.count(opr); }
+  unsigned getLatency(OperatorType *opr) { return latency.lookup(opr); }
+
+  void setStartTime(Operation *op, unsigned val) { startTime[op] = val; }
+
+  //===--------------------------------------------------------------------===//
+  // Client access to properties: specify problem, retrieve solution
+  //===--------------------------------------------------------------------===//
+public:
+  void setAssociatedOperatorType(Operation *op, OperatorType *opr) {
+    assocOpr[op] = opr;
+  }
+
+  void setLatency(OperatorType *opr, unsigned val) { latency[opr] = val; }
+
+  bool hasStartTime(Operation *op) { return startTime.count(op); }
+  unsigned getStartTime(Operation *op) { return startTime.lookup(op); }
+
+  //===--------------------------------------------------------------------===//
+  // Hooks to check/verify the different problem components
+  //===--------------------------------------------------------------------===//
+protected:
+  virtual LogicalResult checkOp(Operation *op);
+  virtual LogicalResult checkDep(Dependence *dep);
+  virtual LogicalResult checkOpr(OperatorType *opr);
+  virtual LogicalResult checkProb();
```

Comment:
Maybe spell out these abbreviations? `Op` is often abbreviated in MLIR projects, so I might not care about that, but the extra characters are worth it to me for the other ones. I know that `Prob` means `Problem` here, but I first read `Probability`, for example.

Reply:
I got rid of the abbreviations everywhere, except in the parameter names -- is that ok?

Reply:
Works for me, thanks! I was mainly concerned about the abbreviated method names.